src/if_statements.py

```diff
@@ -231,17 +231,73 @@ def add_statement(self, statement: Statement) -> None:
     def add_comment(self, contents: str) -> None:
         self.add_statement(SimpleStatement(None, comment=contents))
 
-    def add_if_else(self, if_else: IfElseStatement) -> None:
-        if if_else.if_body.ends_in_jump():
-            # Transform `if (A) { B; return C; } else { D; }`
-            # into `if (A) { B; return C; } D;`,
-            # which reduces indentation to make the output more readable
-            self.statements.append(replace(if_else, else_body=None))
-            if if_else.else_body is not None:
-                self.extend(if_else.else_body)
+    def add_if_else(self, if_else: IfElseStatement, else_jump: bool = False) -> None:
+        if if_else.else_body is not None and not if_else.if_body.ends_in_jump():
+            # Simple case; perform no further rewrites
+            self.statements.append(if_else)
             return
 
-        self.statements.append(if_else)
+        # We now know that we have an IfElseStatement like `if (A) { B; goto C; } else { D; }`
+        # where `D` may be empty. We can rewrite this into `if (A) { B; goto C; } D;`
+        # which reduces indentation to make the output more readable.
```

Comment:
This rewrite only happens at the very end of this function, which is pretty far down. Can you put the middle part in a function? It doesn't need access to the full if_else either, only to `.condition` and `.if_body`, and it may be nicer to have those as individual variables rather than incrementally replace()ing the if_else.

Reply:
I moved the bulk of this fn out to `rewrite_if_ands()`.  Is this sort of what you meant? 
I don't love the required precondition that can't be `assert`ed on (no `else_body` to check if it's empty) -- but moving that check into `rewrite_if_ands()` makes `Body.add_if_else()` practically empty, and loses a bit of that encapsulation.

Reply:
Yeah, your refactoring matches what I was thinking. IMO the fact that the helper doesn't take in an else block means it's clear from the caller that it cannot possibly emit an else, which makes the assert unnecessary. But true, I can see the argument about it potentially being error-prone.